Add a function to build overviews

phargogh opened the pull request:
This PR adds a function to build overviews with a couple of parameters:

* Whether the overviews should be internal or external
* The resampling algorithm to use
* Whether to overwrite existing overviews if they exist
* The overview levels to use, or if the function should just generate a bunch of them.

In addition to having a function to build overviews, I also extended `get_raster_info()` to include a new `overviews` key that lists the dimensions of the overviews.  This is similar to what we see in the output of `gdalinfo`.  I also added another missing `file_list` parameter that was present in `get_raster_info()`'s output but not in the docstring.

Fixes #280

phargogh committed:
Implementing overviews function.  RE:#280

phargogh committed:
Adding a test for overviews and making function importable.

phargogh committed:
Testing internal and external gtiff overviews.

RE:#280

phargogh committed:
Handling (and testing) the case where overviews already exist.

RE:#280

phargogh committed:
Testing path/band tuple format handling.

RE:#280

phargogh committed:
Writing a docstring. RE:#280

phargogh committed:
Overviews are created for the whole raster, not for a single band.

RE:#280

phargogh committed:
Adding list of overviews to ``get_raster_info``.

phargogh committed:
Supporting user-defined overviews levels.

RE:#280

phargogh committed:
Forgot to use a loop variable in a test.

RE:#280

phargogh committed:
Noting change in HISTORY.

RE:#280

dcdenu4 commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -4065,6 +4078,108 @@ def _mult_op(base_array, base_nodata, scale, datatype):
+
+    Args:
+        raster_path (str): A path to a raster on disk and the band index to
+            build overviews for.
```
Minor discrepancy between the text and the input. "and the band index to build ..." but the input is just a string. Should this be `raster_path_band (tuple)`?

phargogh replied:
Whoops!  Forgot to update this message when I changed it from `(path, band)` to just `path`.  I couldn't find an obvious way to tell GDAL to only generate overviews for a single band, so now the function just does it for the whole raster.  Patched in 5678032

dcdenu4 commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -4065,6 +4078,108 @@ def _mult_op(base_array, base_nodata, scale, datatype):
+    Returns:
+        ``None``
+    """
+    # WarpOptions.this is None when an invalid option is passed, and it's a
```
This is a nice tidbit. Are we able to get the possible resample methods from GDAL and provide those in the error message?

phargogh replied:
This ended up being kind of neat!  GDAL doesn't have something like a module-level mapping to get the human-readable types, and the mapping of `gdal.GRA_*` to human-readable names is [local to the `gdal.WarpOptions` parsing function.](https://github.com/OSGeo/gdal/blob/bfe1cfea364a0e185a792db67db7de2fc6aad902/swig/python/osgeo/gdal.py#L854).  I put together something in 7bdd1dbc8a4fc62015c0a737e4712b5022b75708 to get those human-readable names out of `gdal.WarpOptions` (avoiding `ast` inspection in case we can't read the source code like in pyinstaller binaries) and making them available at the module level.  This should make pygeoprocessing resilient to any changes to the set of resample algorithms GDAL offers, so long as they keep `gdal.WarpOptions` up to date!

454dda32c9251191f8f882b1b7a803d594e11693 takes this a step further and uses these human-readable resample algorithms in the docstrings where they are used.

dcdenu4 commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -4065,6 +4078,108 @@ def _mult_op(base_array, base_nodata, scale, datatype):
+            overviews_progress.last_progress_report = time.time()
+    overviews_progress.last_progress_report = time.time()
+
+    raster_path = raster_path
```
Is this necessary or handling a local scope case?

phargogh replied:
It is not necessary!  This is left over from a find/replace and is removed in 5678032.

dcdenu4 reviewed (changes requested):
Thanks @phargogh, I just had a few questions / comments.

phargogh committed:
Cleanup of docstring and unnecessary variable (re)definition.

RE:#280

phargogh committed:
Defining known resample algorithms as a module constant.

GDAL's WarpOptions parsing function (``osgeo.gdal.WarpOptions``) has a
local dict defining a mapping of gdal.GRA_* constants to the appropriate
human-readable strings. The only ways to get at this mapping is to
either parse the source code with ``ast`` or to derive the
human-readable mappings by calling ``gdal.WarpOptions()`` multiple times
and getting the output.  Since we can't always rely on having the source
code available (like in pyinstaller), I went with the black-box method.
It appears to work fine and should be resilient to any changes to GDAL's
selection or naming of resample algorithms.

RE:#280

phargogh committed:
Consolidating human-readable list of resample algos.

Using in docstrings throughout the module.

RE:#280

phargogh committed:
Merge branch 'main' of github.com:natcap/pygeoprocessing into feature/280-overviews

Conflicts:
	HISTORY.rst

phargogh committed:
Noting new function name in HISTORY.

RE:#280

dcdenu4 commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -80,6 +81,23 @@ def __init__(self, missing_values, raster_path, value_map):
     gdal.GDT_CFloat64: numpy.complex64,
 }
 
+# GDAL's python API recognizes certain strings but the only way to retrieve
+# those strings is to do this conversion of gdalconst.GRA_* types to the
+# human-readable labels via gdal.WarpOptions like so.
+_GDAL_WARP_ALGORITHMS = []
+for _warp_algo in (_attrname for _attrname in dir(gdalconst)
+                   if _attrname.startswith('GRA_')):
+    # Appends ['-r', 'near'] to _GDAL_WARP_ALGORITHMS if _warp_algo is
+    # gdalconst.GRA_NearestNeighbor.  See gdal.WarpOptions for the dict
+    # defining this mapping.
+    gdal.WarpOptions(options=_GDAL_WARP_ALGORITHMS,
+                     resampleAlg=getattr(gdalconst, _warp_algo))
```
Very cool!

dcdenu4 reviewed (approved):
Thanks @phargogh! That's a really neat solution for the resampling methods. Thanks for looking into the weeds to pull that out.

It looks like there is a conflict with `History.rst` but I'll go ahead and approve now.

phargogh committed:
Merge branch 'main' into feature/280-overviews

phargogh committed:
Fixing one more outdated comment.  RE:#280

phargogh committed:
Including valid resmple algorithms in exceptions.

RE:#280

phargogh commented:
Thanks @dcdenu4 !  I made two additional changes beyond the merge conflict in order to fix a comment that was made out of date by recent changes and also to include the warp algorithms in the text of exceptions.  So I figure that might deserve a glance over/re-review.

dcdenu4 reviewed (approved):
Awesome, thanks @phargogh !
